### DISCUSSION TITLE
关于OSR与NCW的对比

### DISCUSSION THREAD

**Discussion by njuzhou** [CefView/QCefView#152] [Q&A] [answered]
因为项目网页资源较多，加载性能想更加优化，因此这次尝试了构建NCW模式，在此过程中有一些结论与疑问希望与大家分享

1.两种模式的判断方式
分别使用两种对应库后，希望能够在代码中检验对应模式的库是否使用正确。目前看能够查看代码QCefView中onBrowserWindowCreated是否生效来判断，但不清楚这种判断方式是否准确，或者是否有其他更直接获得此时模式的方式。

2.效能上的提升
在当前使用场景下，我并不是游戏，视频，或者3d效果资源，仅是网页要素及js资源过多。此场景下NCW相比OSR性能提升约20%，即原需要2s加载时间，目前大概1.6s。不确定这种提升是否与大家相同。

3.使用上某些问题
使用NCW模式后，demo可以正常运行，但替换到项目工程后，就始终出现cef对象无法显示的问题，也就是在widget对象展示后，屏幕依然无反应，在OSR模式下工程正常运行，不清楚在使用上，OSR与NCW是否存在差异，有什么要注意的地方

**Comment by tishion** [comment]
> 1.两种模式的判断方式
分别使用两种对应库后，希望能够在代码中检验对应模式的库是否使用正确。目前看能够查看代码QCefView中onBrowserWindowCreated是否生效来判断，但不清楚这种判断方式是否准确，或者是否有其他更直接获得此时模式的方式。

看来有不少用户希望把NCW和OSR模式合并在一起，通过config来实现控制。

**Reply by njuzhou** [reply]
如果能够这样，确实从工程构建和配置查询上会方便些

**Comment by tishion** [comment]
> 2.效能上的提升
在当前使用场景下，我并不是游戏，视频，或者3d效果资源，仅是网页要素及js资源过多。此场景下NCW相比OSR性能提升约20%，即原需要2s加载时间，目前大概1.6s。不确定这种提升是否与大家相同。

NCW和OSR性能的差别在渲染，而不在资源的获取，所以你这里的数据测量方法可能有点问题。

**Reply by njuzhou** [reply]
懂了，感谢！
另外请问如果在资源获取性能上，与浏览器有差异，是有什么启动项配置可以优化吗

**Comment by tishion** [comment]
> 3.使用上某些问题
使用NCW模式后，demo可以正常运行，但替换到项目工程后，就始终出现cef对象无法显示的问题，也就是在widget对象展示后，屏幕依然无反应，在OSR模式下工程正常运行，不清楚在使用上，OSR与NCW是否存在差异，有什么要注意的地方

这个不应该出现，NCW在windows和macos上应该没有这么严重的问题。可以给出更多详细信息一起看看。

**Reply by njuzhou** [reply]
这个可能和我们项目结构相关，只要这两者没有明显使用上的差异，那我就好好排查本身系统设计，感觉像是和父对象属性有关。

**Reply by njuzhou** [reply]
可以确定和父对象相关，具体什么差别还要定位

**Reply by le0zh** [reply]
@tishion NCW 模式下遇到了类似的问题，可以稳定复现的是加了下面两个配置后（无边框+阴影窗口需求）

```
setWindowFlags(Qt::FramelessWindowHint);
setAttribute(Qt::WA_TranslucentBackground);
```

QCefView 无法显示，通过 remote devtools 可以确定 web 资源已经加载成功。

**Reply by tishion** [reply]
这个问题就属于Qt的问题了，不是QCefView能解决的。QCefView不是无法显示，而是QCefView本身就没有任何绘制，所以你设置了WA_TranslucentBackground属性后，QCefView部分就被当成没有绘制的背景区域，然后变成透明，这个时候即使浮在上面的native cef窗口能正常显示，鼠标事件也无法被捕获。

**Reply by le0zh** [reply]
那 NCW 模式没法用在「无边框+阴影」窗口里了么 😢  ，补充一下，FramelessWindowHint 和 WA_TranslucentBackground 同时设置后，才会有问题。单独设置 WA_TranslucentBackground  是可以正常展示的。

**Reply by le0zh** [reply]
> 那 NCW 模式没法用在「无边框+阴影」窗口里了么 😢 ，补充一下，FramelessWindowHint 和 WA_TranslucentBackground 同时设置后，才会有问题。单独设置 WA_TranslucentBackground 是可以正常展示的。

经过调研还是能用，就是麻烦些。

**Reply by tishion** [reply]
有一个解决方案，不知道可行不可行，还没实现完。

**Reply by tishion** [reply]
https://github.com/CefView/QCefView/commit/b03f1b2471b6b3879cfb7d1260938e4cb752154a
这个提交解决了这个问题。

**Reply by njuzhou** [reply]
为大佬效率点赞

**Reply by le0zh** [reply]
@tishion 大佬，在新提交的基础上，发现加了 `config.addCommandLineSwitch("disable-gpu");`  的配置后，也会无法展示，好奇这个和 GPU 加速有什么关系。

**Reply by tishion** [reply]
新版本，OSR和NCW合并了，需要通过Config来配置，默认是OSR模式，你看看是不是没有禁用OSR

**Reply by le0zh** [reply]
最新 main 分支的代码，复现步骤：

1. 修改下面两行配置：

![image](https://user-images.githubusercontent.com/3633972/192205075-c6d1c984-a3d8-4d39-906f-9e7fca628ad5.png)

2. 放开下面两行代码：

![image](https://user-images.githubusercontent.com/3633972/192205271-31d317e3-fea0-45e3-be4d-1d6d10827984.png)

**Reply by tishion** [reply]
refere to:https://www.magpcss.org/ceforum/viewtopic.php?f=6&t=11953

新版本的CEF必须依靠GPU渲染合成，如果你觉得进程太多就加in-proess-gpu

**Reply by njuzhou** [reply]
各位使用新的test加载都正常吗，我使用新的切换NCW模式，设置Frameless和Translucent，却依然无法加载
![F%((P50RUID)W(I G4NZ9JD](https://user-images.githubusercontent.com/25097261/197501578-c4b84ca8-5a68-4f8d-b070-20aad66996fc.png)
![8}XHVF1JO03) 3KC E5(05J](https://user-images.githubusercontent.com/25097261/197501600-f04aa80c-208c-496c-b23d-c3c06ac882ba.png)
![~9~ )MVKQ`NV9_@AP96~)(6](https://user-images.githubusercontent.com/25097261/197501637-04df9708-f1c0-4834-acf4-11aa5b9c10ef.png)

**Reply by tishion** [reply]
![image](https://user-images.githubusercontent.com/5869938/197798651-2ee2cc50-b971-4087-97c0-e667cd9b4952.png)

使用Qt 6.2.4，NCW，Frameless，Translucent

**Reply by njuzhou** [reply]
我使用的Qt5.15.2+VS2019，我再检查一下，感谢大佬

**Reply by sdcarterchen** [reply]
> 我使用的Qt5.15.2+VS2019，我再检查一下，感谢大佬

解决了吗？我的和你环境一样，也是无法渲染出来

**Reply by sdcarterchen** [reply]
> 那 NCW 模式没法用在「无边框+阴影」窗口里了么 😢 ，补充一下，FramelessWindowHint 和 WA_TranslucentBackground 同时设置后，才会有问题。单独设置 WA_TranslucentBackground 是可以正常展示的。

最后解决了吗？

**Reply by tishion** [reply, marked as answer]
Qt 6.4
![image](https://user-images.githubusercontent.com/5869938/233848067-d660711a-534c-47d9-9fc3-b163d2ab1789.png)

**Reply by tishion** [reply]
都已经说了需要Qt版本支持

**Reply by sdcarterchen** [reply]
> 都已经说了需要Qt版本支持

感谢。是我没开GPU，后来在回复中看到了。
